src/nerdbank-streams/src/Channel.ts

```diff
@@ -0,0 +1,135 @@
+import { IDisposableObservable } from "./IDisposableObservable";
+import { MultiplexingStreamClass } from "./MultiplexingStream";
+
+export abstract class Channel implements IDisposableObservable {
+    public duplex: Duplex;
+    public acceptance: Promise<void>;
+    public completion: Promise<void>;
+    private _isDisposed: boolean = false;
+
+    get isDisposed(): boolean {
+        return this._isDisposed;
+    }
+
+    public dispose() {
+        this._isDisposed = true;
+    }
+}
+
+// tslint:disable-next-line:max-classes-per-file
+export class ChannelClass extends Channel {
+    public readonly id: number;
+    private _duplex: Duplex;
+    private readonly _acceptance = new Deferred<void>();
+    private readonly _completion = new Deferred<void>();
+
+    constructor(
+        private multiplexingStream: MultiplexingStreamClass,
+        id: number,
+        public name: string,
+        private offeredByThem: boolean,
+        private options: ChannelOptions) {
+
+        super();
+        const self = this;
+        this.id = id;
+        this._duplex = new Duplex({
+            async write(chunk, encoding, callback) {
+                let error;
+                try {
+                    const payload = Buffer.from(chunk);
+
+                    const header = new FrameHeader();
+                    header.code = ControlCode.Content;
+                    header.channelId = id;
+                    header.framePayloadLength = payload.length;
+                    await multiplexingStream.sendFrameAsync(header, payload);
+                } catch (err) {
+                    error = err;
+                }
+
+                if (callback) {
+                    callback(error);
+                }
+            },
+
+            async final(cb?: (err?: any) => void) {
+                let error;
+                try {
+                    await multiplexingStream.onChannelWritingCompleted(self);
+                } catch (err) {
+                    error = err;
+                }
+
+                if (cb) {
+                    cb(error);
+                }
+            },
+
+            read(size) {
+                // Nothing to do here since data is pushed to us.
+            },
+        });
+    }
+
+    get duplex(): Duplex {
+        return this._duplex;
+    }
+
+    get acceptance(): Promise<void> {
+        return this._acceptance.promise;
+    }
+
+    get acceptanceIsCompleted() {
+        return this._acceptance.isCompleted;
+    }
+
+    get isAccepted() {
+        return this._acceptance.isCompleted;
+    }
+
+    get isRejectedOrCanceled() {
+        return this._acceptance.isRejected;
+    }
+
+    get completion(): Promise<void> {
+        return this._completion.promise;
+    }
+
+    public tryAcceptOffer(options?: ChannelOptions): boolean {
+        if (this._acceptance.resolve()) {
+            return true;
+        }
+
+        return false;
+    }
+
+    public onAccepted() {
```

Comment:
for functions that return nothing, you would usually do : void to make it clear, or in this case, whatever the return of resolve() is (boolean?)